File: baekjoon/1463.py
# X가 3으로 나누어 떨어지면, 3으로 나눈다.
# X가 2로 나누어 떨어지면, 2로 나눈다.
# 1을 뺀다.

import sys
input = sys.stdin.readline

number = int(input())


dp = [0]*(number+1)
for i in range(2, number+1):
    dp[i] = dp[i-1]+1
    # if로 해놔야 %3, %2 둘 다 체크함
    if i % 3 == 0:
        dp[i] = min(dp[i], dp[i//3]+1)
    if i % 2 == 0:
        dp[i] = min(dp[i], dp[i//2]+1)

print(dp[number])

File: javascript/1463.js
const fs = require("fs");
const io = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(io).toString();

const num = Number(input);

const DP = Array(num + 1).fill(0);

DP[1] = 0;
DP[2] = 1;

for (let i = 3; i <= num; i++) {
  DP[i] = DP[i - 1] + 1;
  if (i % 2 === 0) DP[i] = Math.min(DP[i], DP[i / 2] + 1);
  if (i % 3 === 0) DP[i] = Math.min(DP[i], DP[i / 3] + 1);
}

console.log(DP[num]);
